ext/semian/semset.c

```diff
@@ -0,0 +1,69 @@
+#include <semset.h>
+
+const char *SEMINDEX_STRING[] = {
+    FOREACH_SEMINDEX(GENERATE_STRING)
+};
+
+void
+raise_semian_syscall_error(const char *syscall, int error_num)
+{
+  rb_raise(eSyscall, "%s failed, errno: %d (%s)", syscall, error_num, strerror(error_num));
+}
+
+key_t
+generate_sem_set_key(const char *name)
+{
+  char semset_size_key[20];
+  char *uniq_id_str;
+
+  // It is necessary for the cardinatily of the semaphore set to be part of the key
+  // or else sem_get will complain that we have requested an incorrect number of sems
+  // for the desired key, and have changed the number of semaphores for a given key
+  sprintf(semset_size_key, "_NUM_SEMS_%d", SI_NUM_SEMAPHORES);
```

Comment:
This is a bit of a hack, adding the cardinality of the semaphore set to the key before hashing it.

This allows for an upgrade path, increasing the number of semaphores. 

Note that we never actually clean up the semaphore set of the old cardinality, but we could manually if we chose to. Rebooting would also release them.